src/woodbury_pd_mat.jl

```diff
@@ -42,10 +42,13 @@ WoodburyPDMat{T, TA, TD, TS}(vals...) where {T, TA, TD, TS} = WoodburyPDMat(vals
 
 PDMats.dim(W::WoodburyPDMat) = size(W.A, 1)
 
-# Convesion method. Primarily useful for testing purposes.
+# Conversion method. Primarily useful for testing purposes.
 Base.Matrix(W::WoodburyPDMat) = W.A * W.D * W.A' + W.S
+Base.Array(W::WoodburyPDMat) = Matrix(W)
 
-Base.getindex(W::WoodburyPDMat, inds...) = getindex(Matrix(W), inds...)
+function Base.getindex(W::WoodburyPDMat, i1::Integer, i2::Integer)
```

Comment:
AFAICT this will break for `w[i, :]`? We should probably test a few different calls.

We should the backup method since we probably don't want to spend time generalising the faster implementation?

Also a link to the issue would be helpful to see why the new dispatch exists.

Reply:
> Just to understand: previously the string would iterate over all the indexes, which would each construct a full Matrix, and then return the value. Now, instead of constructing the full matrix, we just construct the right element.

Yep, exactly.

> This seems sensible, and I wonder whether it takes roughly the same time as `string(Matrix(woodbury))`?

So, `string(m)` doesn't necessarily attempt to access every element of the matrix, since there is some cutoff width & height `M`. So `string(Matrix(w))` is `O(N^2)`, whereas `string(w)` is (now) `O(min(N, M)^2)`.

For large `N` (where speed is arguably most important), then `string(w)` will win. For `N < M`, constructing a dense matrix first will definitely be faster, due to better use of cache / wide registers. We'd have to profile to find for what `N` it becomes faster to use the proposed `getindex`, but I think (for now) this is unimportant.

(NB, I'm ignoring a linear factor in terms of the low-rank dimension, but I think for the analysis above this is fine)

> AFAICT this will break for w[i, :]? We should probably test a few different calls.

I've added more tests! Actually it worked fine for all other indexing styles, albeit inefficiently, since they all (eventually) delegate to `getindex(::WoodburyPDMat, ::Int, ::Int)`.

I've also added a few optimised dispatches. I don't think we need to keep the old implementation around, unless there's anything else I've missed?

> Also a link to the issue would be helpful to see why the new dispatch exists.

Sure! (although I'd argue that the new implementation is the "obvious" approach...)

Reply:
Nice, thanks for the explanation. I would add a range index test as well, and that then covers the most common cases.